base_layer/mmr/src/merkle_mountain_range.rs

```diff
@@ -137,35 +137,48 @@ where
         );
         Ok(peaks
             .into_iter()
-            .map(|i| self.hashes.get_or_panic(i))
+            .map(|i| {
+                self.hashes
+                    .get(i)
+                    .unwrap()
+                    .expect("find_peaks returned invalid indexes")
+            })
             .fold(hasher, |hasher, h| hasher.chain(h)))
     }
 
     /// Push a new element into the MMR. Computes new related peaks at the same time if applicable.
     /// Returns the new length of the merkle mountain range (the number of all nodes, not just leaf nodes).
-    pub fn push(&mut self, hash: &Hash) -> Result<usize, MerkleMountainRangeError> {
+    pub fn push(&mut self, hash: Hash) -> Result<usize, MerkleMountainRangeError> {
         if self.is_empty()? {
-            return self.push_hash(hash.clone());
+            return self.push_hash(hash);
         }
+
         let mut pos = self.len()?;
         let (peak_map, height) = peak_map_height(pos);
         if height != 0 {
             return Err(MerkleMountainRangeError::CorruptDataStructure);
         }
+
         self.push_hash(hash.clone())?;
+        let mut last_hash = hash;
         // hash with all immediately preceding peaks, as indicated by peak map
         let mut peak = 1;
         while (peak_map & peak) != 0 {
+            // left_sibling can never be out of bounds (>= len) because peak can never be 0
             let left_sibling = pos + 1 - 2 * peak;
-            let left_hash = &self.hashes.get_or_panic(left_sibling);
+            debug_assert!(
+                left_sibling < pos,
+                "left_sibling was greater than the reported number of elements contained in the impl of ArrayLike"
+            );
+            let left_hash = self
+                .hashes
+                .get(left_sibling)
+                .map_err(MerkleMountainRangeError::backend_error)?
+                .unwrap();
             peak *= 2;
             pos += 1;
-            let hash_count = self
```

Comment:
👁️  Please check my logic here. 

last_hash (@ hash_count - 1) should always be the previous hash we pushed to the MMR

Reply:
I think this is the real optimisation.

Removing get_or_panic is a slowdown. I'd leave it in, or else `impl Index` so that you can do a direct lookup (which is basically the same thing under the hood for ArrayLikes that aren't backed by Index)